src/control/server/ctl_storage_rpc.go

```diff
@@ -43,87 +42,88 @@ const (
-	resp = &ctlpb.PrepareNvmeResp{}
-	msg := "Storage Prepare NVMe"
+// doNvmePrepare issues prepare request and returns response.
+func (c *StorageControlService) doNvmePrepare(req *ctlpb.PrepareNvmeReq) *ctlpb.PrepareNvmeResp {
 	_, err := c.NvmePrepare(bdev.PrepareRequest{
 		HugePageCount: int(req.GetNrhugepages()),
 		TargetUser:    req.GetTargetuser(),
 		PCIWhitelist:  req.GetPciwhitelist(),
 		ResetOnly:     req.GetReset_(),
 	})
 
-	if err != nil {
-		resp.State = newState(c.log, ctlpb.ResponseStatus_CTL_ERR_NVME, err.Error(), "", msg)
-		return
-	}
+	pnr := new(ctlpb.PrepareNvmeResp)
+	pnr.State = newResponseState(err, ctlpb.ResponseStatus_CTL_ERR_NVME, "")
 
-	resp.State = newState(c.log, ctlpb.ResponseStatus_CTL_SUCCESS, "", "", msg)
-	return
+	return pnr
 }
 
-func (c *StorageControlService) doScmPrepare(pbReq *ctlpb.PrepareScmReq) (pbResp *ctlpb.PrepareScmResp) {
-	pbResp = &ctlpb.PrepareScmResp{}
-	msg := "Storage Prepare SCM"
+// newPrepareScmResp sets protobuf SCM prepare response with results.
+func newPrepareScmResp(inResp *scm.PrepareResponse, inErr error) (*ctlpb.PrepareScmResp, error) {
+	var info string
+	var err error
+	outResp := new(ctlpb.PrepareScmResp)
 
-	scmState, err := c.GetScmState()
-	if err != nil {
-		pbResp.State = newState(c.log, ctlpb.ResponseStatus_CTL_ERR_SCM, err.Error(), "", msg)
-		return
-	}
-	c.log.Debugf("SCM state before prep: %s", scmState)
+	defer func() {
+		if err != nil {
+			return
+		}
+		outResp.State = newResponseState(inErr, ctlpb.ResponseStatus_CTL_ERR_SCM, info)
+	}()
 
-	resp, err := c.ScmPrepare(scm.PrepareRequest{Reset: pbReq.Reset_})
-	if err != nil {
-		pbResp.State = newState(c.log, ctlpb.ResponseStatus_CTL_ERR_SCM, err.Error(), "", msg)
-		return
+	if inErr != nil {
+		return outResp, nil
 	}
 
-	info := ""
-	if resp.RebootRequired {
-		info = scm.MsgScmRebootRequired
+	if inResp.RebootRequired {
+		info = scm.MsgRebootRequired
+		outResp.Rebootrequired = true
 	}
-	pbResp.Rebootrequired = resp.RebootRequired
 
-	pbResp.Namespaces = make(proto.ScmNamespaces, 0, len(resp.Namespaces))
-	if err := (*proto.ScmNamespaces)(&pbResp.Namespaces).FromNative(resp.Namespaces); err != nil {
-		pbResp.State = newState(c.log, ctlpb.ResponseStatus_CTL_ERR_SCM, err.Error(), "", msg)
-		return
+	outResp.Namespaces = make(proto.ScmNamespaces, 0, len(inResp.Namespaces))
+	err = (*proto.ScmNamespaces)(&outResp.Namespaces).FromNative(inResp.Namespaces)
+
+	return outResp, err
+}
+
+func (c *StorageControlService) doScmPrepare(pbReq *ctlpb.PrepareScmReq) (*ctlpb.PrepareScmResp, error) {
+	scmState, err := c.GetScmState()
+	if err != nil {
+		return newPrepareScmResp(nil, err)
 	}
-	pbResp.State = newState(c.log, ctlpb.ResponseStatus_CTL_SUCCESS, "", info, msg)
+	c.log.Debugf("SCM state before prep: %s", scmState)
 
-	return
+	return newPrepareScmResp(c.ScmPrepare(scm.PrepareRequest{Reset: pbReq.Reset_}))
```

Comment:
I know we do this a lot already, but I'm not really crazy about nesting nontrivial function calls.

Reply:
you are against function chaining? what's the particular reason?
would you prefer that we call the individual functions, create local variables for the results and pass them in turn to the functions in sequence?
that seems to me a bit of a C-language style restriction that isn't usually applied in languages like Go, but it's obviously a highly subject style preference. @mjmac thoughts?

Reply:
Yeah, it's very subjective. For me it's a readability thing. It takes a bit more effort to figure out where the real action is happening. I don't have a problem chaining functions that are doing simple things like getting and mutating data, but c.ScmPrepare is a major state-changing call. I haven't had to run a debugger on our Go code yet, but when running gdb on C/C++ code, things can get confusing fast with a couple layers of chained functions.